### ISSUE TITLE
Suppress runtime exceptions thrown from Gauge

### ISSUE THREAD

**Issue by ryantenney** [dropwizard/metrics#642]

**Comment by tkowalcz** [comment]
Hi,

this is something we ran into recently. I see that you do not want to handle exceptions from metric classes, but currently this exception will not be reported anywhere and GraphiteReporter will silently stop sending metrics.

Are you going to merge this change or at least log the exception?

**Comment by dbaggott** [comment]
@tkowalcz (and future readers), it's worth noting that the unpleasant "all reporters silently die in the face of any runtime exception" bug present in version 3.0.2 has been addressed in version 3.1.0 and onward.

`ScheduledReporter.start()` [now correctly handles any exception](https://github.com/dropwizard/metrics/blob/v3.1.0/metrics-core/src/main/java/com/codahale/metrics/ScheduledReporter.java#L118) that bubbles up to it: the exception will be logged and subsequent reporting events are not suppressed.

So, as the code currently stands, a `RuntimeException` from a custom `Gauge` (or any other `Metric`) will break reporting for that cycle but reporting will continue to be attempted on the regular schedule.  More importantly, a clear error will be logged which should help people understand what's going on quickly.  Thank you to whoever made that change!

If the Metric runtime exception is intermittent, you'll still get data for reporting cycles where the exception doesn't happen.

All of that being said, @ryantenney, I love the Metrics library but I have to say there's really not a lot of great choices for developers in this scenario and I think there's a pretty reasonable case to be made that the library should support handling `RuntimeException`s from `Metric`s.  Consider this interface:

```java
public interface Gauge<T> extends Metric {
    T getValue();
}
```

For example, let's say it's a `Gauge<Integer>` that reports on the number of "some undesirable event" and your code encounters an exception, what's the correct return value?

You definitely don't want to return 0 and risk falsely representing that there are no such events.  Nor do you want to return some "magic" number like Integer.MAX_VALUE as its misleading and, in that example, blows out the y-axis on any downstream graphs.  A "magic" number like -1 is a little better but it's still problematic since it can appear awfully close to 0 on a graph.  In short, in order to not obscure downstream graphs by blowing out the axis, your "magic" number would have to be tuned to the range of the actual data!  It just not a workable solution.

As the Metrics library code currently stands, returning *null* is the *only* thing to do if you don't want to create a magic error value for your Gauge.  Like I said, not a lot of good choices here!  The worst of it is that you're forced to rely on undocumented behavior of the Metrics library that it handles nulls (the GraphiteReporter does and Dropwizard's /metrics do but I have no idea how other reporters behave).  It's just not a great situation to have to report an exceptional condition with a null nor is it, in my opinion, reasonable or desirable to actually require all consumers of the data to do null checks.

Honestly, throwing a  `RuntimeException` from the Gauge strikes me as the "right" thing to do as (given the API) *this level of the code* isn't actually equipped to know how to handle the exceptional condition.

Finally, I'll add that the serializer that renders `Gauge`s for Dropwizard's /metrics API explicitly handles `RuntimeException`s\.

Thoughts?

**Comment by JakePi3** [comment]
I'm seeing some flakiness with the code in question.  Getting an exception deep within the Jackson serialization to JSON and it's not bubbling up at all.  The thread is dying silently.  I'm using 3.1.2.

Two things to think about:

ExecutorService returns a Future which captures the exception and does not actually bubble up until .get() is called on the Future - at which point it throws an ExecutionException.

https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Future.html#get()

Assuming ScheduledExecutorService functions in the same way.

Secondly, it may not be enough to capture RuntimeException. This is one of those rare cases where you want to catch on Throwable.  Otherwise any Error thrown will kill the thread silently.
